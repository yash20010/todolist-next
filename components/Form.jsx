import styles from '../styles/Home.module.css'
import axios from 'axios'

const url = 'http://localhost:3000/api/task'

const Form = ({ tasks, setTasks, task, setTask }) => {
  const handleChange = ({ currentTarget: input }) => {
    input.value === ''
      ? setTask({ task: '' })
      : setTask((prev) => ({ ...prev, task: input.value }))
  }

  const addTask = async (e) => {
    e.preventDefault()
    try {
      if (task._id) {
        const { data } = await axios.put(url + '/' + task._id, {
          task: task.task,
        })
        const originalTasks = [...tasks]
        const index = originalTasks.findIndex((t) => t._id === task._id)
        originalTasks[index] = data.data
        setTasks(originalTasks)
        setTask({ task: '' })
        console.log(data.message)
      } else {
        const { data } = await axios.post(url, task)
        setTasks((prev) => [...prev, data.data])
        setTask({ task: '' })
        console.log(data.message)
      }
    } catch (error) {
      console.log(error)
    }
  }

  return (
    <form onSubmit={addTask} className={styles.form_container}>
      <input
        className={styles.input}
        type="text"
        placeholder="Task to be done..."
        onChange={handleChange}
        value={task.task}
        data-testid="input"
      />
      <button
        type="submit"
        className={styles.submit_btn}
        disabled={!task.task}
        data-testid="button"
      >
        {task._id ? 'Update' : 'Add'}
      </button>
    </form>
  )
}

export default Form
