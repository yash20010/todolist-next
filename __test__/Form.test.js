import React from 'react'
import ReactDOM from 'react-dom'
import Form from '../components/Form.jsx'
import { render, fireEvent, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'

afterEach(cleanup)

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Form task={{ task: '' }} />, div)
})

it('renders the placeholder text of the input field correctly', () => {
  const { getByTestId } = render(<Form task={{ task: '' }} />)
  expect(getByTestId('input').placeholder).toBe('Task to be done...')
})

it('renders the button text correctly while adding a todo', () => {
  const { getByTestId } = render(<Form task={{ task: '' }} />)
  expect(getByTestId('button')).toHaveTextContent('Add')
})

it('button is disabled on render, before user starts typing', () => {
  const { getByTestId } = render(<Form task={{ task: '' }} />)
  expect(getByTestId('button')).toBeDisabled()
})

it('button is enabled when user starts typing', () => {
  const { getByTestId } = render(<Form task={{ task: 'hello' }} />)
  expect(getByTestId('button')).not.toBeDisabled()
})

it('renders the button correctly during updating a todo', () => {
  const { getByTestId } = render(
    <Form
      task={{
        _id: 1234,
        task: 'hello',
        completed: false,
        createdAt: new Date(),
      }}
    />
  )
  expect(getByTestId('button')).toHaveTextContent('Update')
})

it('matches snapshot', () => {
  const tree = renderer.create(<Form task={{ task: '' }} />).toJSON()
  expect(tree).toMatchSnapshot()
})
