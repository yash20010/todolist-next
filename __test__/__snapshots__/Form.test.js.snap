// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className="form_container"
  onSubmit={[Function]}
>
  <input
    className="input"
    data-testid="input"
    onChange={[Function]}
    placeholder="Task to be done..."
    type="text"
    value=""
  />
  <button
    className="submit_btn"
    data-testid="button"
    disabled={true}
    type="submit"
  >
    Add
  </button>
</form>
`;
